### ISSUE TITLE
Setup script, .env file support in launchSettings.json

### ISSUE THREAD

**Issue by markm77** [dotnet/runtime#69379]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

In production we use Linux .env files to configure large numbers of environment variables. For Windows development, we can use a JSON file (user secrets) for this purpose but this is a completely incompatible format so difficult to do things like maintain a common subset of variables or temporarily test somewhere variables from another environment. With WSL2, we can create something like a "local prod" environment but can't easily run it with a specified .env file. Note that .env files allow environment variables to be specified outside of the project repo (like user secrets). Docker also makes extensive use of .env files which should really be specified in launchSettings (there is currently a setting at .csproj level) so that they can be added only when appropriate (.e.g. perhaps not in "Development" if user secrets are used in that case).

Sometimes instead of using .env files we'd like to run a startup script which might e.g. create an environment based on parameters or obtain secrets from a cloud source. We may not wish to put this in our code in order to ensure same behaviour regardless of how an environment is set up.

The current options of directly-specified launchSettings.json environment variables (very bad for open source repos) or use of user secrets (in JSON) seem very limiting.

### Describe the solution you'd like

Support of .env files and setup scripts (e.g. bash or PowerShell) in launchSettings for Windows, WSL2 and Docker launch scenarios.

**Comment by davidfowl** [comment]
Dupe of https://github.com/dotnet/runtime/issues/49667

**Comment by davidfowl** [comment]
Oh wait, this isn't about runtime, this is about tooling and launch scenarios. 

cc @vijayrkn

**Comment by AaronRobinsonMSFT** [comment]
@jkoritzinsky and I have discussed this before. We support [.env files](https://smartmob-rfc.readthedocs.io/en/latest/2-dotenv.html) in [`corerun`](https://github.com/dotnet/runtime/blob/main/docs/workflow/testing/using-corerun.md) and I can say that it has improved my inner dev loop a great deal. Adding support for `dotnet`/apphost would be interesting to spec out and consider the benefits.

/cc @richlander

**Comment by markm77** [comment]
Thanks for the comments. Just to make clear the most immediate problem is that I am maintaining duplicate `secrets.json` and `.env` files to cover different launch scenarios which is a pain. I want to standardise on `.env`. I also want to be able to use setup scripts to allow customisation of env vars depending on e.g. user, permissions, environment, etc and to ensure certain env vars are not stored locally but only cloud-sourced as needed.

**Comment by markm77** [comment]
Another thing to mention if you're thinking about this. It's useful to be able to easily convert between appsettings.json and .env file formats. This is because it is easier to edit appsettings.json files when there is heavy nesting or you want to "paste in" JSON. But .env is needed to configure env vars. Of course we can all write our own scripts for this but having a standardised converter utility might help keep things standardised (e.g. use of "__" separator etc).

(Someone reading this will be wondering why not just use appsettings.json files directly? The answer is many of us want to use env vars because they keeps configuration out of the source repo and uncoupled from the application and give flexibility for deployment scenarios. : ) )

**Comment by markm77** [comment]
Thanks very much for comments @vijayrkn and appreciate you looking into this.

> This ask is mostly about the issue linked https://github.com/dotnet/runtime/issues/49667 (Having a default env configuration provider similar to user secrets).

A configuration provider would certainly be very useful. The reason I suggested allowing direct use of an .env file from launchSettings.json is that this ensures when we try to replicate prod setup on WSL2 (for example) the env vars will be identical and the same "configuration gathering" code will run (helpful for debug). It also matches the fact that this is where other env vars are  usually set (a .env file is really just env vars). Env vars of a process can also often be inspected etc. But if what we get is a configuration provider, that will be still be a great improvement on today.

**Comment by markm77** [comment]
I should also add that if you give us setup scripts, that would anyway cover the `.env` file request as we can then e.g. create a script to `source` the .env file into a WSL2 Bash shell before application start-up.

>  Is this profile level/project level or repo level?

I think it's really important to be able to have multiple launch profiles in a launchSettings.json that each behave differently. They can operate in very different environments (e.g. Docker, WSL2, Windows).

**Comment by vijayrkn** [comment]
> The reason I suggested allowing direct use of an .env file from launchSettings.json is that this ensures when we try to replicate prod setup on WSL2 (for example) the env vars will be identical and the same.

I see. So, if I were to visualize your ask, you are looking for some way to define in the launchProfile to load the environment variables directly from an .env file (https://docs.docker.com/compose/env-file/#syntax-rules) rather than defining it in the environmentVariables section. 

```
    "WebApplication1": {
      "commandName": "Project",
      "dotnetRunMessages": true,
      "launchBrowser": true,
      "applicationUrl": "https://localhost:7214;http://localhost:5214",
      "envFileEnabled": true,
      "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "WSL2"
      }
    }
```

Also adding @BillHiebert (who is the primary dev for the tooling area) to gather his thoughts as well.

**Comment by markm77** [comment]
Yes, exactly. It seems to me that loading from an .env file should be the exact equivalent of listing the env vars one by one in the launch profile. It just has the benefit of allowing them to be defined and managed externally (and out-of-repo).

It would also be great if specifying more than one .env file per launch profile were supported as some of us like to split env vars into groups. : )

**Comment by vijayrkn** [comment]
> Yes, exactly. It seems to me that loading from an .env file should be the exact equivalent of listing the env vars one by one in the launch profile. 

Makes sense. Thanks for the clarification.

**Comment by richlander** [comment]
I saw this comment: https://github.com/dotnet/runtime/issues/69379#issuecomment-1127218536. I like the idea of having support in the host for this scenario but I haven't heard much call for it. Starting with support in tools seems like a fine place to start. 

However, it gets more weird if `launchSettings.json` references specific `.env` files since one is intended for source control and the other (often) not.

**Comment by markm77** [comment]
> However, it gets more weird if launchSettings.json references specific .env files since one is intended for source control and the other (often) not.

Thanks for the comment. I'm not wedded to lauchSettings.json as long as the use of .env files can be controlled in different launch scenarios. For instance, I'd be fine with storing .env files in the UserSecrets folder (actually I do this already). And perhaps the filename could be used to indicate environment as for appsettings.json (e.g. myenv.{Envrionment}.env)?

**Comment by markm77** [comment]
> I like the idea of having support in the host for this scenario but I haven't heard much call for it.

The driver for me is that I'm building an open-source *application*. (I guess there is a lot of open-source software but perhaps fewer fully-featured, back-end applications.) I don't want to hard-code configuration gathering patterns with configuration providers etc in my app like I would with a proprietary app because I want to give the user deployment flexibility and freedom. Therefore .env files/env vars are the best way to handle configuration and secrets because the user can generate them or cloud-source them as they wish. And configuration management is then completely decoupled from the application.

**Comment by markm77** [comment]
BTW the above example could also allow supporting `.sh`, `.ps1` etc files (setup scripts!) as `environmentFiles` where some platform rules applied (e.g. `.sh` files not supported on Windows launch profiles).

That would look a pretty nice design to me!

**Comment by markm77** [comment]
Another way to handle the case where external files may not be present would be to have `"environmentFiles"` and `"optionalEnvironmentFiles"` properties where the latter are not required but used if available.

**Comment by richlander** [comment]
Cool. I'll leave the design converation to @vijayrkn.
